src/apiutils.jl

```diff
@@ -17,17 +17,19 @@ end
 ###################################
 # vector mode function evaluation #
 ###################################
-
-@generated function dualize(::Type{T}, x::SArray{S,V,D,N}) where {T,S,V,D,N}
+@generated function dualize(::Type{T}, x::Union{FieldVector, SArray}) where {T}
+    N = length(x)
+    V = eltype(x)
```

Comment:
The type constraint on `x` allows this (and the other similar changes in this PR) to technically work, since we can probably assume that no upstream code will re-overload `length` or `eltype` for `Union{FieldVector, SArray}`. 

However, it's a bit sketchy...it makes it easier for bugs to accidentally be introduced in the future via seemingly innocent changes (e.g. loosening the type constraint on `x`).

Would it be hard to just pull these out of the type parameter directly instead?

Reply:
It seems like calling `length` and `eltype` makes this more straightforward to extend to additional types. @zsunberg and I decided to retrieve `N` and `V` via `length` and `eltype` because it was already quite complicated to try to extract them directly from the type parameters.

I've tried a few different variations on this:
```julia
@generated function dualize2(::Type{T}, x::Union{FieldVector{N,V}, SArray{S,V,D,N}}) where {T,S,V,D,N}
```
Everything I've tried fails when called on an instance of `FieldVector`, although most of the approaches seem to work on `SVector`. I'm not 100% sure what you mean by "loosen the type constraint", but it seems like that would cause an even bigger headache if we want to extract the parameters this way.
 
If we tried to extract type parameters directly within the function body, we'd need a separate case for each type (because `V` and `N` are in different places for each type):
```julia
N,V = supertype(typeof(x)).parameters              # <: FieldVector
S,V,D,N = typeof(x).parameters             # SArray
```

Either of the above ways requires more customization than simply calling `length` and `eltype`. I think this way is more stable. It should work for vector types for which `length` and `eltype` are defined.

Another option is to have two different function definitions in every case where we need `V` and `N`.

Reply:
> I think this way is more stable. It should work for vector types for which length and eltype are defined.

I think you might misunderstand my concern, which is my bad - I didn't explain the issue directly enough. If this were normal code, then of course I'd be fine with `length` and `eltype` getting used here. 

However, this code is being called in an `@generated` thunk, and in general, calling overloadable methods inside of `@generated` function thunks is something that should be done with care, because downstream code can cause the generator to be "invalidated" in ways the compiler can't easily predict. Such code can cause world-age errors or silently incorrect method lookups, and are very hard for end-users to fix/debug/understand because they can depend on precompilation settings, code load order, etc. Thus, we want to be very careful about potentially introducing these kinds of bugs.

In this case, since `FieldVector` is an abstract type, so I could subtype it and overload `length` and `eltype` on that subtype. For the sake of example, let's call this subtype `MyFieldVectorSubtype`.  If a user loaded ForwardDiff, called `dualize` on some other `<:FieldVector`, then loaded the code that defines `FieldVectorSubtype`, then called `dualize` on an instance of `MyFieldVectorSubtype`, the compiler could possibly throw a world-age error (most likely), silently call the wrong methods, run into some other compilation error, etc.

While this may seem like a niche class of bugs, they can crop up in practice quite regularly; for one ForwardDiff-related instance of such a bug, see https://github.com/JuliaArrays/StaticArrays.jl/issues/316.

By the way, I really appreciate this PR, and am excited for this feature!

Reply:
Ah ok, yeah, we didn't understand the specific concerns with @generated

What about
```julia
@generated function dualize(::Type{T}, x::Union{FieldVector{N,V}, SArray{<:Any,V,<:Any,N}}) where {T,V,N}
```
?

Reply:
(we didn't know about the <:Any syntax in the type parameters)

Reply:
If that doesn't work, I think two dualize functions would be a little simpler than _length and _eltype

Reply:
@jrevels, thanks for clarifying! This makes more sense. We'll incorporate the changes you suggested and get back to you.